pkg/prom/cleaner.go

```diff
@@ -0,0 +1,258 @@
+// Override default age after which abandoned WALs can be removed
+func WithCleanerMinAge(d time.Duration) WALCleanerOpts {
+	return func(c *WALCleaner) {
+		if d > 0 {
+			c.minAge = d
+		}
+	}
+}
+
+// Override default period for checking for abandoned WALs
+func WithCleanerPeriod(d time.Duration) WALCleanerOpts {
+	return func(c *WALCleaner) {
+		if d > 0 {
+			c.ticker = time.NewTicker(d)
+		}
+	}
+}
+
+//
+type WALCleaner struct {
+	logger          log.Logger
+	instanceManager instance.Manager
+	walDirectory    string
+	walOperations   walOperations
+	minAge          time.Duration
+	ticker          *time.Ticker
+	done            chan bool
+}
+
+// Create a new cleaner that looks for abandoned WALs in the given directory and
+// removes them if they haven't been modified in over minAge. Starts a goroutine
+// to periodically run WALCleaner.CleanupStorage in a loop
+func NewWALCleaner(logger log.Logger, manager instance.Manager, walDirectory string, opts ...WALCleanerOpts) *WALCleaner {
+	c := &WALCleaner{
+		logger:          log.With(logger, "component", "cleaner"),
+		instanceManager: manager,
+		walDirectory:    filepath.Clean(walDirectory),
+		walOperations:   &walOperationsDefault{},
+		minAge:          DefaultCleanupAge,
+		done:            make(chan bool),
+	}
+
+	for _, opt := range opts {
+		opt(c)
+	}
+
+	// If the caller hasn't set a ticker, create one with the default period
+	if c.ticker == nil {
+		c.ticker = time.NewTicker(DefaultCleanupPeriod)
+	}
+
+	go c.run()
+	return c
+}
+
+// Get storage directories used for each ManagedInstance
+func (c *WALCleaner) getManagedStorage(instances map[string]instance.ManagedInstance) map[string]bool {
+	out := make(map[string]bool)
+
+	for _, inst := range instances {
+		out[inst.StorageDirectory()] = true
+	}
+
+	return out
+}
+
+// Get all Storage directories under walDirectory
+func (c *WALCleaner) getAllStorage() []string {
+	var out []string
+
+	_ = filepath.Walk(c.walDirectory, func(p string, info os.FileInfo, err error) error {
+		if err != nil {
+			if os.IsNotExist(err) {
+				// The root WAL directory doesn't exist. Maybe this Agent isn't responsible for any
+				// instances yet. Log at debug since this isn't a big deal. We'll just try to crawl
+				// the direction again on the next periodic run.
+				level.Debug(c.logger).Log("msg", "WAL storage path does not exist", "path", p, err)
+			} else {
+				// Just log any errors traversing the WAL directory. This will potentially result
+				// in a WAL (that has incorrect permissions or some similar problem) not being cleaned
+				// up. This is  better than preventing *all* other WALs from being cleaned up.
+				discoveryError.WithLabelValues(p).Inc()
+				level.Warn(c.logger).Log("msg", "unable to traverse WAL storage path", "path", p, "err", err)
+			}
+		} else if info.IsDir() && filepath.Dir(p) == c.walDirectory {
+			// Single level below the root are instance storage directories (including WALs)
+			out = append(out, p)
+		}
```

Comment:
Can we flatten this out into one if tree?

```go
if os.IsNotExist(err) { 
   // ...
} else if err != nil {
  // .. 
} else if info.IsDir() && filepath.Dir(p) == c.walDirectory {
  // ... 
}
return nil 
```

If you feel it looks better you can try a switch too:

```go
switch {
case os.IsNotExist(err):
case err != nil:
case info.IsDir() && filepath.Dir(p) == c.walDirectory:
}
```

Reply:
Yep. Any preference on if/else vs switch?